File: lemonadeChange.go
package main

import(
    "fmt"
    //"strings"
    //"sort"
)

var testMod=false;

func test() {
}

func lemonadeChange(bills []int) bool {
    log:=map[int]int{5:0,10:0}
    for _,bill:=range bills {
        switch bill {
        case 5:
            log[5]++
        case 10:
            if log[5]==0 {
                return false
            }
            log[5]--
            log[10]++
        case 20:
            if log[5]==0 {
                return false
            }
            if log[10]!=0 {
                log[10]--
                log[5]--
            } else if log[5]>=3 {
                log[5]-=3
            } else {
                return false
            }
        }
    }
    return true
}

func main() {
    if testMod {
        test()
    }
    inp:=[]int{10,10}
    fmt.Println(lemonadeChange(inp))
}

File: lemonadeChange.py
from typing import List


class Solution:
    def lemonadeChange(self, bills: List[int]) -> bool:
        res5, res10 = 0, 0

        for b in bills:
            if b == 5:
                res5 += 1
            elif b == 10:
                res10 += 1

                if res5 > 0:
                    res5 -= 1
                else:
                    return False
            else:
                if res10 > 0:
                    res10 -= 1

                    if res5 > 0:
                        res5 -= 1
                    else:
                        return False
                elif res5 > 2:
                    res5 -= 3
                else:
                    return False

        return True


bills = [5, 5, 5, 10, 20]
bills = [5, 5, 10, 10, 20]
print(Solution().lemonadeChange(bills))
